File: lib/models/Trip.ts
import mongoose, { Document, Schema } from "mongoose";

export interface ILocation {
  type: "campus" | "off-campus";
  campusLocation?: string;
  offCampusAddress?: string;
}

export interface ITrip extends Document {
  travelerId: mongoose.Types.ObjectId;
  fromLocation: ILocation;
  toLocation: ILocation;
  departureTime: Date;
  transportMethod: string;
  maxDeliveries: number;
  currentDeliveries: number;
  pricePerDelivery: number; // Price in Ghana Cedi (GHC)
  isRecurring: boolean;
  status: "active" | "completed" | "cancelled";
  matchedRequests: mongoose.Types.ObjectId[];
  joinedUsers: mongoose.Types.ObjectId[];
  createdAt: Date;
  updatedAt: Date;
}

const LocationSchema = new Schema<ILocation>(
  {
    type: {
      type: String,
      enum: ["campus", "off-campus"],
      required: true,
    },
    campusLocation: {
      type: String,
      required: function () {
        return this.type === "campus";
      },
    },
    offCampusAddress: {
      type: String,
      required: function () {
        return this.type === "off-campus";
      },
    },
  },
  { _id: false }
);

const TripSchema = new Schema<ITrip>(
  {
    travelerId: {
      type: Schema.Types.ObjectId,
      ref: "User",
      required: [true, "Traveler ID is required"],
    },
    fromLocation: {
      type: LocationSchema,
      required: [true, "From location is required"],
    },
    toLocation: {
      type: LocationSchema,
      required: [true, "To location is required"],
    },
    departureTime: {
      type: Date,
      required: [true, "Departure time is required"],
    },
    transportMethod: {
      type: String,
      required: [true, "Transport method is required"],
      enum: ["Car", "Motorcycle", "Bicycle", "Walking", "Public Transport"],
    },
    maxDeliveries: {
      type: Number,
      required: [true, "Maximum deliveries is required"],
      min: [1, "Maximum deliveries must be at least 1"],
      max: [10, "Maximum deliveries cannot exceed 10"],
    },
    currentDeliveries: {
      type: Number,
      default: 0,
      min: 0,
    },
    pricePerDelivery: {
      type: Number,
      required: [true, "Price per delivery is required"],
      min: [0, "Price cannot be negative"],
    },
    isRecurring: {
      type: Boolean,
      default: false,
    },
    status: {
      type: String,
      enum: ["active", "completed", "cancelled"],
      default: "active",
    },
    matchedRequests: [
      {
        type: Schema.Types.ObjectId,
        ref: "DeliveryRequest",
      },
    ],
    joinedUsers: [
      {
        type: Schema.Types.ObjectId,
        ref: "User",
      },
    ],
  },
  {
    timestamps: true,
  }
);

// Indexes for better query performance
TripSchema.index({ travelerId: 1 });
TripSchema.index({ status: 1 });
TripSchema.index({ departureTime: 1 });
TripSchema.index({ "fromLocation.type": 1, "fromLocation.campusLocation": 1 });
TripSchema.index({ "toLocation.type": 1, "toLocation.campusLocation": 1 });

export default mongoose.models.Trip ||
  mongoose.model<ITrip>("Trip", TripSchema);

File: backend/internal/models/trip.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type TripStatus string
type TransportMethod string

const (
	TripActive    TripStatus = "active"
	TripCompleted TripStatus = "completed"
	TripCancelled TripStatus = "cancelled"
)

const (
	TransportCar        TransportMethod = "car"
	TransportMotorcycle TransportMethod = "motorcycle"
	TransportBicycle    TransportMethod = "bicycle"
	TransportWalking    TransportMethod = "walking"
	TransportPublic     TransportMethod = "public_transport"
)

type Trip struct {
	ID                uuid.UUID       `json:"id" db:"id"`
	TravelerID        uuid.UUID       `json:"travelerId" db:"traveler_id"`
	FromLocation      string          `json:"fromLocation" db:"from_location"`
	ToLocation        string          `json:"toLocation" db:"to_location"`
	DepartureTime     time.Time       `json:"departureTime" db:"departure_time"`
	TransportMethod   TransportMethod `json:"transportMethod" db:"transport_method"`
	MaxDeliveries     int             `json:"maxDeliveries" db:"max_deliveries"`
	CurrentDeliveries int             `json:"currentDeliveries" db:"current_deliveries"`
	PricePerDelivery  float64         `json:"pricePerDelivery" db:"price_per_delivery"`
	IsRecurring       bool            `json:"isRecurring" db:"is_recurring"`
	Status            TripStatus      `json:"status" db:"status"`
	Description       *string         `json:"description" db:"description"`
	ContactInfo       *string         `json:"contactInfo" db:"contact_info"`
	CreatedAt         time.Time       `json:"createdAt" db:"created_at"`
	UpdatedAt         time.Time       `json:"updatedAt" db:"updated_at"`

	// Populated fields
	Traveler        *User             `json:"traveler,omitempty"`
	TravelerName    string            `json:"travelerName,omitempty"`
	JoinedUsers     []User            `json:"joinedUsers,omitempty"`
	MatchedRequests []DeliveryRequest `json:"matchedRequests,omitempty"`
}

type CreateTripRequest struct {
	FromLocation     string          `json:"fromLocation" validate:"required"`
	ToLocation       string          `json:"toLocation" validate:"required"`
	DepartureDate    string          `json:"departureDate" validate:"required"`
	DepartureTime    string          `json:"departureTime" validate:"required"`
	AvailableSeats   int             `json:"availableSeats" validate:"required,gt=0"`
	PricePerDelivery float64         `json:"pricePerDelivery" validate:"required,gt=0"`
	VehicleType      TransportMethod `json:"vehicleType" validate:"required"`
	Description      *string         `json:"description"`
	ContactInfo      *string         `json:"contactInfo"`
}

type JoinTripRequest struct {
	TripID uuid.UUID `json:"tripId" validate:"required"`
}

type LeaveTripRequest struct {
	TripID uuid.UUID `json:"tripId" validate:"required"`
}

// Junction table for trip participants
type TripParticipant struct {
	TripID uuid.UUID `db:"trip_id"`
	UserID uuid.UUID `db:"user_id"`
}

// Junction table for matched delivery requests
type TripDeliveryRequest struct {
	TripID            uuid.UUID `db:"trip_id"`
	DeliveryRequestID uuid.UUID `db:"delivery_request_id"`
}
